# Data Understanding — 

inspect tables, columns, grain, keys, relationships.

In [40]:
import pandas as pd 
import numpy as np

In [42]:
accounts = pd.read_csv("/Users/falesh/Data Science/Customer Retention Intelligence/saas-subscription-analytics/accounts.csv")

In [44]:
print(accounts.shape)

(1200, 7)


In [47]:
accounts.info()

<class 'pandas.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   account_id      1200 non-null   int64
 1   company_name    1200 non-null   str  
 2   industry        1200 non-null   str  
 3   country         1200 non-null   str  
 4   employee_count  1200 non-null   int64
 5   signup_date     1200 non-null   str  
 6   plan            1200 non-null   str  
dtypes: int64(2), str(5)
memory usage: 65.8 KB


In [94]:
accounts.columns

Index(['account_id', 'company_name', 'industry', 'country', 'employee_count',
       'signup_date', 'plan'],
      dtype='str')

In [46]:
accounts.head()

,account_id,company_name,industry,country,employee_count,signup_date,plan
0,221118,Xenon Analytics,Logistics,Japan,3,2024-07-22 17:11:47,Starter
1,697547,TriggerPoint,Software,Brazil,75,2025-02-08 13:10:44,Growth
2,753247,MeshWorks,Manufacturing,United Kingdom,30,2023-11-17 15:15:33,Growth
3,931016,Catalyst CRM,Retail,Germany,58,2023-06-13 07:14:32,Growth
4,254108,Junction Cloud,Financial Services,Canada,11,2024-04-18 18:14:32,Starter


In [49]:
accounts['account_id'].nunique()

1200

accounts
Primary key candidate: account_id
Rows: 1,200

In [50]:
accounts.duplicated().sum()

np.int64(0)

In [51]:
accounts['plan'].value_counts()

plan
Starter       554
Growth        473
Business      159
Enterprise     14
Name: count, dtype: int64

In [52]:
users = pd.read_csv('/Users/falesh/Data Science/Customer Retention Intelligence/saas-subscription-analytics/users.csv')

In [55]:
users.shape
users.info()
users.head()

<class 'pandas.DataFrame'>
RangeIndex: 21884 entries, 0 to 21883
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   user_id     21884 non-null  int64
 1   account_id  21884 non-null  int64
 2   full_name   21884 non-null  str  
 3   role        21884 non-null  str  
 4   email       21884 non-null  str  
dtypes: int64(2), str(3)
memory usage: 855.0 KB


,user_id,account_id,full_name,role,email
0,1,574882,Edward Edwards,member,eedwards@funnelsoftgroup1.com
1,2,574882,Aiden Morris,member,amorris@funnelsoftgroup1.com
2,3,574882,Lakshmi Nair,member,lnair@funnelsoftgroup1.com
3,4,574882,Marisol Lopez,admin,mlopez@funnelsoftgroup1.com
4,5,574882,Feng Cao,member,fcao@funnelsoftgroup1.com


In [59]:
users.nunique()

user_id       21884
account_id     1200
full_name      5489
role              3
email         21884
dtype: int64

users
Primary key candidate: user_id
account_id: foreign-key candidate
Rows: 21,884

In [61]:
users['role'].value_counts()

role
member    13515
viewer     4374
admin      3995
Name: count, dtype: int64

In [63]:
users.duplicated().sum() # checks duplicate rows

np.int64(0)

In [65]:
subscriptions = pd.read_csv('/Users/falesh/Data Science/Customer Retention Intelligence/saas-subscription-analytics/subscriptions.csv')

In [68]:
subscriptions.shape
subscriptions.info()
subscriptions.head()

<class 'pandas.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   subscription_id  1200 non-null   int64  
 1   account_id       1200 non-null   int64  
 2   seats            1200 non-null   int64  
 3   status           1200 non-null   str    
 4   mrr              1200 non-null   float64
 5   started_on       1200 non-null   str    
 6   ended_on         86 non-null     str    
dtypes: float64(1), int64(3), str(3)
memory usage: 65.8 KB


,subscription_id,account_id,seats,status,mrr,started_on,ended_on
0,500001,574882,6,active,72.0,2025-02-20 11:17:51,NaN
1,500002,561622,3,active,36.0,2024-06-09 16:53:00,NaN
2,500003,952954,19,churned,456.0,2023-08-03 20:58:36,2024-04-04 20:58:36
3,500004,656283,35,active,840.0,2024-01-27 17:16:01,NaN
4,500005,541269,9,churned,216.0,2024-06-24 23:24:02,2025-03-21 23:24:02


In [70]:
subscriptions.nunique()

subscription_id    1200
account_id         1200
seats               127
status                4
mrr                 153
started_on         1200
ended_on             86
dtype: int64

In [71]:
subscriptions["status"].value_counts()

status
active      1030
churned       86
trialing      46
past_due      38
Name: count, dtype: int64

In [72]:
subscriptions.duplicated().sum()

np.int64(0)

In [73]:
invoices = pd.read_csv('/Users/falesh/Data Science/Customer Retention Intelligence/saas-subscription-analytics/invoices.csv')

In [75]:
invoices.shape
invoices.info()
invoices.head()

<class 'pandas.DataFrame'>
RangeIndex: 14500 entries, 0 to 14499
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   invoice_id    14500 non-null  int64  
 1   account_id    14500 non-null  int64  
 2   invoice_date  14500 non-null  str    
 3   amount        14500 non-null  float64
 4   status        14500 non-null  str    
dtypes: float64(1), int64(2), str(2)
memory usage: 566.5 KB


,invoice_id,account_id,invoice_date,amount,status
0,5439659,877086,2025-07-20 11:35:55,573.25,paid
1,6345938,466636,2025-05-03 09:14:19,125.17,paid
2,5942838,346678,2025-05-12 23:24:32,384.25,paid
3,4660126,790364,2025-10-18 11:31:31,178.09,paid
4,4894712,315012,2025-11-14 16:53:30,3878.60,paid


In [76]:
invoices.nunique()

invoice_id      14500
account_id       1200
invoice_date    14495
amount          12010
status              4
dtype: int64

In [77]:
invoices['status'].value_counts()

status
paid             13031
open               834
void               327
uncollectible      308
Name: count, dtype: int64

In [79]:
invoices.duplicated().sum()

np.int64(0)

In [80]:
support_tickets = pd.read_csv('/Users/falesh/Data Science/Customer Retention Intelligence/saas-subscription-analytics/support_tickets.csv')

In [81]:
support_tickets.shape

(5600, 8)

In [83]:
support_tickets.info()
support_tickets.head()

<class 'pandas.DataFrame'>
RangeIndex: 5600 entries, 0 to 5599
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ticket_id           5600 non-null   int64  
 1   account_id          5600 non-null   int64  
 2   opened_at           5600 non-null   str    
 3   priority            5600 non-null   str    
 4   category            5600 non-null   str    
 5   satisfaction_score  5129 non-null   float64
 6   resolution_hours    5129 non-null   float64
 7   resolved_at         5129 non-null   str    
dtypes: float64(2), int64(2), str(4)
memory usage: 350.1 KB


,ticket_id,account_id,opened_at,priority,category,satisfaction_score,resolution_hours,resolved_at
0,563761,466834,2025-04-21 10:28:33,normal,Billing,NaN,NaN,NaN
1,364489,509193,2025-10-29 20:26:40,normal,Billing,1.0,41.3,2025-10-31 13:44:40
2,1274,263552,2025-10-24 11:15:00,normal,Integrations,4.0,59.8,2025-10-26 23:03:00
3,671186,412093,2025-09-10 20:04:41,normal,Bug Report,NaN,NaN,NaN
4,184373,150144,2025-12-26 16:11:30,normal,Onboarding,5.0,16.5,2025-12-27 08:41:30


In [86]:
support_tickets.nunique()


ticket_id             5600
account_id            1186
opened_at             5600
priority                 4
category                 6
satisfaction_score       5
resolution_hours      1193
resolved_at           5129
dtype: int64

In [88]:
support_tickets['priority'].value_counts()

priority
normal    2533
low       1601
high      1077
urgent     389
Name: count, dtype: int64

In [89]:
support_tickets['category'].value_counts()

category
Integrations       1274
Bug Report         1140
Billing             978
Onboarding          749
Performance         730
Feature Request     729
Name: count, dtype: int64

In [85]:
support_tickets.duplicated().sum()

np.int64(0)

In [90]:
users[~users["account_id"].isin(accounts["account_id"])]["account_id"].nunique()

0

In [91]:
subscriptions[~subscriptions["account_id"].isin(accounts["account_id"])]["account_id"].nunique()

0

In [92]:
support_tickets[~support_tickets['account_id'].isin(accounts['account_id'])]['account_id'].nunique()

0

In [93]:
invoices[~invoices["account_id"].isin(accounts["account_id"])]["account_id"].nunique()

0